### ISSUE TITLE
Custom visitor

### ISSUE THREAD

**Issue by jjimenezg93** [flang-compiler/f18#428]
Hi,
We're trying to create our own visitor to create a customized output where we print, among other things, the location of the given node in the source (its provenance), the type of statement or its name (if it's for example a variable or a procedure). This is common functionality, but access to this information varies between classes.

The templated `Walk()` overloads already provide an appropriate tree visiting mechanism, so the only "viable" approach we've found so far is to replicate the class `ParseTreeDumper`, creating as many `Pre()` and `Post()` overloads as necessary.

The problem is that some "nodes" have provenance, others might not (I'm not completely sure) and, most importantly, **the access** to this provenance **varies** from class to class. For example, we can get the provenance of a `NamedConstant` through its attribute `v,` which is of type `Name`:

```c++
#define WRAPPER_CLASS_BOILERPLATE(classname, type) \
  BOILERPLATE(classname); \
  classname(type &&x) : v(std::move(x)) {} \
  using WrapperTrait = std::true_type; \
  type v

#define WRAPPER_CLASS(classname, type) \
  struct classname { \
    WRAPPER_CLASS_BOILERPLATE(classname, type); \
  }
// ...
WRAPPER_CLASS(NamedConstant, Name);
```

However, another class, such as an `IfThenStmt`, does not have a `v` attribute and the access to its provenance is different.

Assuming all of them have a provenance, which seems so since most are defined either by a `WRAPPER_CLASS` or by a `TUPLE_CLASS` with a hierarchy which ultimately has a `Name` somewhere, we could simply generate as many `Pre()` and `Post()` as necessary to handle all cases (it might even be possible to [group by access](https://stackoverflow.com/a/16000226/3459662)), but I don't think that's a good solution, although I can't see a better one right now.

Is there any **better**/**simpler** approach to this problem already implemented or even in consideration?

Thanks in advance!

**Comment by tskeith** [comment]
Not every node has provenance information available. It's only available through parser::CharBlock objects. Several kinds of nodes have `CharBlock source;` members and those are the ones you can map back to file names and character positions.

I'm not sure this fully answers your question, so please reopen if you need more information.

**Comment by jjimenezg93** [comment]
Hi @tskeith,
I'm aware of that, our problem is not in provenances itself, but how to reach to them from the tree visitation. I might not have explained myself properly, please let me try again.

What **we want** to achieve is:

Given the following F90 program

```fortran
PROGRAM Fortran90

TYPE S
  INTEGER a
  INTEGER(KIND=8) b
END TYPE S

CONTAINS
! ...
FUNCTION func0(x)
! ...
END FUNCTION func0
END PROGRAM Fortran90
```

We want to print the tree with this information of all the nodes (we're aware that some nodes might lack provenance):

```
MainProgram 0x561122986798 <f.f90:1:1, f.f90:62:21>
|-ProgramStmt 0x561122986d50 <f.f90:1:1, f.f90:62:21> fortran90
|-SpecificationPart 0x561122987098 <f.f90:2:1, f.f90:7:1>
| |-ImplicitPart 0x561122986d30 <<invalid sloc>>
| `-DerivedTypeDef 0x5611229876d8 <f.f90:3:1, f.f90:6:10> s
|   |-DataComponentDefStmt 0x5611229e4730 <f.f90:4:3, f.f90:4:11> a 'i32'
|   `-DataComponentDefStmt 0x5611229e4790 <f.f90:5:3, f.f90:5:19> b 'i64'
| InternalSubprogramPart 0x561122987348  <f.f90:8:1, f.f90:62:21>
| |-ContainsStmt 0x561122987160 <f.f90:8:1, f.f90:8:8> 
| |-FunctionSubprogram 0x5611229e4bc0 <f.f90:10:1, f.f90:15:18> func0 'float (i32)'
...
```

As I mentioned, I've seen that many of the structs defined in `parse-tree.h` end up having a `Name&` at some point in the hierarchy, from which we could obtain `CharBlock source` and `Symbol* symbol`. Since we've already successfully retrieved provenances from symbols for variables and functions, I think we could work with that if we manage to obtain these `Name`s.

**Current approach**: I've been trying to replicate `ParseTreeDumper`, and that's what I've come up for now (please have in mind it's a Work In Progress PoC):

```
class CustomDumper {
// NODE and NODE_NAME macros boilerplate, just copy&pasted from ParseTreeDumper
...
    template <typename T> bool Pre(const T& x) {
    if constexpr (WrapperTrait<T>) {
      os_ << FindName(x.v) << std::endl;
    }
    return true;
  }
...
template <typename T> std::string FindName(const T& x) {
    if constexpr (WrapperTrait<T>) {
      return FindName(x.v);
    } else if constexpr (UnionTrait<T>) {
      return FindName(x.u);
    } else {
      return "None";
    }
  }

  template <typename T> std::string FindName(const std::string& x) {
    return x;
  }

  template <typename T> std::string FindName(const Fortran::parser::Name& x) {
    return x.symbol->name().ToString();
  }

  template <typename T> std::string FindName(const std::list<T>& x) {
    std::string ret("");
    for (auto& elem : x) {
      ret += FindName(elem) + " | ";
    }
    return ret;
  }

  template <typename... T> std::string FindName(const std::variant<T...>& x) {
    auto& a = std::get<0>(x); // WORKS
    // auto& a = std::get<x.index()>(x); // DOESN'T WORK (runtime)
    return "";
  }
}
```

I know there's an attribute `v` in all classes defined with `WRAPPER_CLASS`, so I can call `FindName(x.v)` when the `WrapperTrait<T>` is true. Then there are multiple template overloads of `FindName()`, some of them are more generic and simply call another overload which will hopefully be more specific.

Let's take `ProgramUnit` (`parse-tree.h`) as an example:
```
struct ProgramUnit {
  UNION_CLASS_BOILERPLATE(ProgramUnit);
  std::variant<common::Indirection<MainProgram>,
      common::Indirection<FunctionSubprogram>,
      common::Indirection<SubroutineSubprogram>, common::Indirection<Module>,
      common::Indirection<Submodule>, common::Indirection<BlockData>>
      u;
};
```

Ok, so for this and any other `std::variant<T...>`, we want the implementation of this overload to simply pass *whatever value* is currently held inside the variant to *whatever other overload* of `FindName` (for a `common::indirection` of `MainProgram`, `FunctionSubprogram`, `SubroutineSubprogram`, etc.) in which this fits. As an example, when the variant `u` inside `ProgramUnit` is sent while processing the first node, it would end up in an overload of `FindName` which can handle a `common::Indirection<MainProgram>`.

**However**, we can't get the current value held by the `std::variant`, since template must be resolved at compile time and `x.index()` is runtime information in this case.

We'd like to know if there's another approach I haven't found in F18 which would make our life much easier :D

We are truly looking forward to achieve this, so please help us!

Thank you very much in advance.

**Comment by tskeith** [comment]
I think you need something like this to handle the generic variant case:
```
template <typename... T> std::string FindName(const std::variant<T...> &x) {
  return std::visit([&](const auto &y) { return FindName(y); });
}
```

**Comment by jjimenezg93** [comment]
Indeed, it looks like this way I can handle the generic `std::variant` case, although please note that there's an `x` missing as the second parameter to `std::visit`.

Thanks for your help.
